lib/stm32/fdcan.c

```diff
@@ -0,0 +1,891 @@
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <libopencm3/stm32/fdcan.h>
+#include <libopencm3/stm32/rcc.h>
+
+
+/* Timeout for FDCAN_CCCR register INIT bit to accept set value.
+ * This timeout is required because FDCAN uses two different clocks
+ * feeding two different portions of block. There can be slight delay
+ * based on how clocks are set up. While amount of FDCAN_clk /
+ * FDCAN_pclk combinations is high and clock speeds may vary a lot,
+ * following value has been choosen as sane default.
+ */
+#define FDCAN_CCCR_INIT_TIMEOUT		0x0000FFFF
+
+static int fdcan_cccr_init_cfg(uint32_t canport, bool set);
+static int fdcan_get_fill_rxfifo(uint32_t canport, uint8_t fifo_id, 
+		unsigned * get_index, unsigned * pending_frames);
+static struct FDCAN_MessageRAM * fdcan_get_msgram_addr(uint32_t canport);
+static int fdcan_get_free_txbuf(uint32_t canport);
+static uint32_t fdcan_length_to_dlc(uint8_t length);
+static uint8_t fdcan_dlc_to_length(uint32_t dlc);
+
+/** @ingroup fdcan_file */
+/**@{
+ * */
+
+/** @brief FDCAN Reset
+ *
+ * This will reset the FDCAN peripheral. All registers will be placed in
+ * the reset state. Note that it is not possible to reset FDCAN per-CAN
+ */
+void fdcan_reset(void)
+{
+	rcc_periph_reset_pulse(RST_FDCAN);
+}
+
+
+/* @brief FDCAN Init
+ * Initialize the selected CAN peripheral block. This function will switch CAN block
+ * into initialization mode and set crucial configuration flags and timing registers.
+ * CAN block is then left in initialization mode in order to 
+ * 
+ * @param[in] canport Unsigend int32. CAN register base address @ref can_reg_base.
+ * @param[in] brs bool. Enable bitrate switching during Tx (only avaiable if FDCAN operation is also enabled)
+ * @param[in] fdoe Enable FDCAN operation
+ * @param[in] nart Disable automatic retransmission
+ * @param[in] rflm Turn on receive FIFO locked mode, otherwize FIFO is in overwrite mode
+ * @param[in] txqm Switch Tx into queue mode, if false then Tx works in FIFO mode 
+ * @param[in] sjw Resynchronization time quanta jump width.
+ * @param[in] ts1 Time segment 1 time quanta width.
+ * @param[in] ts2 Time segment 2 time quanta width.
+ * @param[in] brp Baud rate prescaler.
+ * @param[in] loopback Turn on loopback mode
+ * @param[in] silent Turn on silent (monitor) mode
+ * @returns int FDCAN_E_OK on success, FDCAN_E_TIMEOUT on timeout waiting for INIT bit to flip.
+ */
+int fdcan_init(uint32_t canport, bool brs_enable, bool fd_op_enable, 
+		 bool auto_retr_disable, bool rx_fifo_locked, bool tx_queue_mode, 
```

Comment:
cmon, just add the y in retry?  that doesn't worth trying to save a character over.